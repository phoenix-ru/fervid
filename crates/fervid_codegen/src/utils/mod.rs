// to_ident_or_str

use std::fmt::{Write, Error};

use fervid_core::{FervidAtom, StrOrExpr};
use swc_core::{common::Span, ecma::ast::{ComputedPropName, Ident, EsReserved, IdentName, PropName, Str}};

/// Adapted from SWC Ident::verify_symbol
#[inline]
pub fn is_valid_ident(s: &str) -> bool {
    if s.is_reserved() || s.is_reserved_in_strict_mode(true) || s.is_reserved_in_strict_bind() {
        return false;
    }

    let mut chars = s.chars();

    if let Some(first) = chars.next() {
        if Ident::is_valid_start(first) && chars.all(Ident::is_valid_continue) {
            return true;
        }
    }

    false
}

pub fn str_to_propname(s: &str, span: Span) -> PropName {
    if is_valid_ident(s) {
        PropName::Ident(IdentName { span, sym: s.into() })
    } else {
        PropName::Str(Str {
            span,
            value: s.into(),
            raw: None,
        })
    }
}

pub fn atom_to_propname(sym: FervidAtom, span: Span) -> PropName {
    if is_valid_ident(&sym) {
        PropName::Ident(IdentName { span, sym })
    } else {
        PropName::Str(Str {
            span,
            value: sym,
            raw: None,
        })
    }
}

pub fn str_or_expr_to_propname(str_or_expr: StrOrExpr, span: Span) -> PropName {
    match str_or_expr {
        StrOrExpr::Str(sym) => atom_to_propname(sym, span),
        StrOrExpr::Expr(expr) => PropName::Computed(ComputedPropName {
            span,
            expr,
        }),
    }
}

pub fn to_camelcase(s: &str, buf: &mut impl Write) -> Result<(), Error> {
    for (idx, word) in s.split('-').enumerate() {
        if idx == 0 {
            buf.write_str(word)?;
            continue;
        }

        let first_char = word.chars().next();
        if let Some(ch) = first_char {
            // Uppercase the first char and append to buf
            for ch_component in ch.to_uppercase() {
                buf.write_char(ch_component)?;
            }

            // Push the rest of the word
            buf.write_str(&word[ch.len_utf8()..])?;
        }
    }

    Ok(())
}
