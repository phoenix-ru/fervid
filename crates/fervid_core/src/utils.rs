use swc_core::{common::Span, ecma::ast::{ComputedPropName, EsReserved, Ident, IdentName, PropName, Str}};

use crate::{AttributeOrBinding, FervidAtom, StrOrExpr, VBindDirective};

/// Checks whether the attributes name is the same as `expected_name`
#[inline]
pub fn check_attribute_name(attr: &AttributeOrBinding, expected_name: &str) -> bool {
    matches!(attr,
        AttributeOrBinding::RegularAttribute { name, .. } |
        AttributeOrBinding::VBind(VBindDirective { argument: Some(StrOrExpr::Str(name)), .. })
        if name == expected_name
    )
}

/// Adapted from SWC Ident::verify_symbol
#[inline]
pub fn is_valid_ident(s: &str) -> bool {
    if s.is_reserved() || s.is_reserved_in_strict_mode(true) || s.is_reserved_in_strict_bind() {
        return false;
    }

    is_valid_propname(s)
}

#[inline]
pub fn is_valid_propname(s: &str) -> bool {
    let mut chars = s.chars();

    if let Some(first) = chars.next() {
        if Ident::is_valid_start(first) && chars.all(Ident::is_valid_continue) {
            return true;
        }
    }

    false
}

pub fn str_to_propname(s: &str, span: Span) -> PropName {
    if is_valid_propname(s) {
        PropName::Ident(IdentName { span, sym: s.into() })
    } else {
        PropName::Str(Str {
            span,
            value: s.into(),
            raw: None,
        })
    }
}

pub fn atom_to_propname(sym: FervidAtom, span: Span) -> PropName {
    if is_valid_propname(&sym) {
        PropName::Ident(IdentName { span, sym })
    } else {
        PropName::Str(Str {
            span,
            value: sym,
            raw: None,
        })
    }
}

pub fn str_or_expr_to_propname(str_or_expr: StrOrExpr, span: Span) -> PropName {
    match str_or_expr {
        StrOrExpr::Str(sym) => atom_to_propname(sym, span),
        StrOrExpr::Expr(expr) => PropName::Computed(ComputedPropName {
            span,
            expr,
        }),
    }
}
