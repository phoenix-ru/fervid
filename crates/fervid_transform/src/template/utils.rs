use fervid_core::FervidAtom;
use swc_core::{common::DUMMY_SP, ecma::ast::{ArrowExpr, BindingIdent, BlockStmtOrExpr, Expr, Ident, Pat}};

/// `foo-bar-baz` -> `FooBarBaz`
#[inline]
pub(crate) fn to_pascal_case(raw: &str, out: &mut String) {
    for word in raw.split('-') {
        let first_char = word.chars().next();
        if let Some(ch) = first_char {
            // Uppercase the first char and append to buf
            for ch_component in ch.to_uppercase() {
                out.push(ch_component);
            }

            // Push the rest of the word
            out.push_str(&word[ch.len_utf8()..]);
        }
    }
}

/// `foo-bar-baz` -> `fooBarBaz`
#[inline]
pub(crate) fn to_camel_case(raw: &str, out: &mut String) {
    for (idx, word) in raw.split('-').enumerate() {
        if idx == 0 {
            out.push_str(word);
            continue;
        }

        let first_char = word.chars().next();
        if let Some(ch) = first_char {
            // Uppercase the first char and append to buf
            for ch_component in ch.to_uppercase() {
                out.push(ch_component);
            }

            // Push the rest of the word
            out.push_str(&word[ch.len_utf8()..]);
        }
    }
}

/// Wraps `expr` to `$event => (expr)`
#[inline]
pub fn wrap_in_event_arrow(expr: Box<Expr>) -> Box<Expr> {
    let evt_param = Pat::Ident(BindingIdent {
        id: Ident {
            span: DUMMY_SP,
            ctxt: Default::default(),
            sym: FervidAtom::from("$event"),
            optional: false,
        },
        type_ann: None,
    });

    Box::new(Expr::Arrow(ArrowExpr {
        span: DUMMY_SP,
        ctxt: Default::default(),
        params: vec![evt_param],
        body: Box::new(BlockStmtOrExpr::Expr(expr)),
        is_async: false,
        is_generator: false,
        type_params: None,
        return_type: None,
    }))
}
