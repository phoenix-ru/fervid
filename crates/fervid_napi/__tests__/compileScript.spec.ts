import { describe, test, expect } from 'vitest'
import { assertCode, compile } from './utils'

describe('SFC compile <script setup>', () => {
  // https://github.com/vuejs/core/blob/530d9ec5f69a39246314183d942d37986c01dc46/packages/compiler-sfc/__tests__/compileScript.spec.ts#L16-L52
  test('should expose top level declarations', () => {
    const { content } = compile(`
      <script setup>
      import { x } from './x'
      let a = 1
      const b = 2
      function c() {}
      class d {}
      </script>

      <script>
      import { xx } from './x'
      let aa = 1
      const bb = 2
      function cc() {}
      class dd {}
      </script>
      `)

    expect(content).toMatch(
      `        return {
            get xx () {
                return xx;
            },
            get aa () {
                return aa;
            },
            set aa (v){
                aa = v;
            },
            bb,
            cc,
            dd,
            get x () {
                return x;
            },
            get a () {
                return a;
            },
            set a (v){
                a = v;
            },
            b,
            c,
            d
        };`,
    )

    // expect(bindings).toStrictEqual({
    //   x: BindingTypes.SETUP_MAYBE_REF,
    //   a: BindingTypes.SETUP_LET,
    //   b: BindingTypes.SETUP_CONST,
    //   c: BindingTypes.SETUP_CONST,
    //   d: BindingTypes.SETUP_CONST,
    //   xx: BindingTypes.SETUP_MAYBE_REF,
    //   aa: BindingTypes.SETUP_LET,
    //   bb: BindingTypes.LITERAL_CONST,
    //   cc: BindingTypes.SETUP_CONST,
    //   dd: BindingTypes.SETUP_CONST,
    // })
    assertCode(content)
  })

  // Difference with the original is that in TS `x` and `xx` may be types,
  // and compiler doesn't know without looking at the usage.
  // Since neither are used, they are excluded.
  test('should expose top level declarations w/ ts', () => {
    const { content } = compile(`
      <script setup lang="ts">
      import { x } from './x'
      let a = 1
      const b = 2
      function c() {}
      class d {}
      </script>

      <script lang="ts">
      import { xx } from './x'
      let aa = 1
      const bb = 2
      function cc() {}
      class dd {}
      </script>
      `)

    expect(content).toMatch(
      `        return {
            get aa () {
                return aa;
            },
            set aa (v){
                aa = v;
            },
            bb,
            cc,
            dd,
            get a () {
                return a;
            },
            set a (v){
                a = v;
            },
            b,
            c,
            d
        };`,
    )

    assertCode(content)
  })
})

describe('SFC analyze <script> bindings', () => {
// https://github.com/vuejs/core/blob/272ab9fbdcb1af0535108b9f888e80d612f9171d/packages/compiler-sfc/__tests__/compileScript.spec.ts#L1252-L1306
  describe('auto name inference', () => {
    test('basic', () => {
      const { content } = compile(
        `<script setup>const a = 1</script>
          <template>{{ a }}</template>`,
        {
          filename: 'FooBar.vue',
        },
      )
      expect(content).toMatch(`export default {
    __name: "FooBar"`)
      assertCode(content)
    })

    test('do not overwrite manual name (object)', () => {
      const { content } = compile(
        `<script>
          export default {
            name: 'Baz'
          }
          </script>
          <script setup>const a = 1</script>
          <template>{{ a }}</template>`,
        {
          filename: 'FooBar.vue',
        },
      )
      expect(content).not.toMatch(`name: 'FooBar'`)
      expect(content).toMatch(`name: 'Baz'`)
      assertCode(content)
    })

    test('do not overwrite manual name (call)', () => {
      const { content } = compile(
        `<script>
          import { defineComponent } from 'vue'
          export default defineComponent({
            name: 'Baz'
          })
          </script>
          <script setup>const a = 1</script>
          <template>{{ a }}</template>`,
        {
          filename: 'FooBar.vue',
        },
      )
      expect(content).not.toMatch(`name: 'FooBar'`)
      expect(content).toMatch(`name: 'Baz'`)
      assertCode(content)
    })
  })
})

describe('SFC genDefaultAs', () => {
  test('normal <script> only', () => {
    const { content } = compile(
      `<script>
      export default {}
      </script>`,
      {
        genDefaultAs: '_sfc_',
      },
    )
    expect(content).not.toMatch('export default')
    expect(content).toMatch(`const _sfc_ = {\n    __name: "anonymous"\n}`)
    assertCode(content)
  })

  test('normal <script> w/ cssVars', () => {
    const { content } = compile(
      `<script>
      export default {}
      </script>
      <style>
      .foo { color: v-bind(x) }
      </style>`,
      {
        genDefaultAs: '_sfc_',
      },
    )
    expect(content).not.toMatch('export default')
    expect(content).not.toMatch('__default__')
    expect(content).toMatch(`const _sfc_ = {\n    __name: "anonymous"\n}`)
    assertCode(content)
  })

  test('<script> + <script setup>', () => {
    const { content } = compile(
      `<script>
      export default {}
      </script>
      <script setup>
      const a = 1
      </script>`,
      {
        genDefaultAs: '_sfc_',
      },
    )
    expect(content).not.toMatch('export default')
    expect(content).toMatch(
      // Fervid never produces `Object.assign` because it assumes code downleveling is done by a bundler
      // `const _sfc_ = /*#__PURE__*/Object.assign(__default__`,
      `const _sfc_ = {`,
    )
    assertCode(content)
  })

  test('<script setup> only', () => {
    const { content } = compile(
      `<script setup>
      const a = 1
      </script>`,
      {
        genDefaultAs: '_sfc_',
      },
    )
    expect(content).not.toMatch('export default')
    expect(content).toMatch(`const _sfc_ = {\n    __name: "anonymous",\n    setup`)
    assertCode(content)
  })

  test('<script setup> only w/ ts', () => {
    const { content } = compile(
      `<script setup lang="ts">
      const a = 1
      </script>`,
      {
        genDefaultAs: '_sfc_',
      },
    )
    expect(content).not.toMatch('export default')
    // TODO https://github.com/phoenix-ru/fervid/issues/23
    // expect(content).toMatch(`const _sfc_ = /*#__PURE__*/_defineComponent(`)
    expect(content).toMatch(`const _sfc_ = _defineComponent({`)
    assertCode(content)
  })

  test('<script> + <script setup> w/ ts', () => {
    const { content } = compile(
      `<script lang="ts">
      export default {}
      </script>
      <script setup lang="ts">
      const a = 1
      </script>`,
      {
        genDefaultAs: '_sfc_',
      },
    )
    expect(content).not.toMatch('export default')
    expect(content).toMatch(
      // TODO https://github.com/phoenix-ru/fervid/issues/23
      // There is no need for spreading, because Fervid merges trivial objects
      // `const _sfc_ = /*#__PURE__*/_defineComponent({\n  ...__default__`,
      `const _sfc_ = _defineComponent({\n    __name:`,
    )
    assertCode(content)
  })

  // TODO implement TS-only macros
  // This is tested in Rust side
  // test('binding type for edge cases', () => {
  //   const { bindings } = compile(
  //     `<script setup lang="ts">
  //     import { toRef } from 'vue'
  //     const props = defineProps<{foo: string}>()
  //     const foo = toRef(() => props.foo)
  //     </script>`,
  //   )
  //   expect(bindings).toStrictEqual({
  //     toRef: BindingTypes.SETUP_CONST,
  //     props: BindingTypes.SETUP_REACTIVE_CONST,
  //     foo: BindingTypes.SETUP_REF,
  //   })
  // })

  // describe('parser plugins', () => {
    // Compiler never throws (only during panic)
    // test('import attributes', () => {
    //   const { content } = compile(`
    //     <script setup>
    //     import { foo } from './foo.js' with { type: 'foobar' }
    //     </script>
    //   `)
    //   assertCode(content)

    //   expect(() =>
    //     compile(`
    //   <script setup>
    //     import { foo } from './foo.js' assert { type: 'foobar' }
    //     </script>`),
    //   ).toThrow()
    // })

    // This is not supported
    // test('import attributes (user override for deprecated syntax)', () => {
    //   const { content } = compile(
    //     `
    //     <script setup>
    //     import { foo } from './foo.js' assert { type: 'foobar' }
    //     </script>
    //   `,
    //     {
    //       babelParserPlugins: [
    //         ['importAttributes', { deprecatedAssertSyntax: true }],
    //       ],
    //     },
    //   )
    //   assertCode(content)
    // })
  // })
})
