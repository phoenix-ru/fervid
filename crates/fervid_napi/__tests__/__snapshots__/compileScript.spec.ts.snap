// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SFC analyze <script> bindings > auto name inference > basic 1`] = `
"import { toDisplayString as _toDisplayString } from "vue";
export default {
    __name: "FooBar",
    render (_ctx, _cache, $props, $setup, $data, $options) {
        return _toDisplayString(a);
    },
    setup (__props, { expose: __expose }) {
        __expose();
        const a = 1;
        return {
            a
        };
    }
};
"
`;

exports[`SFC analyze <script> bindings > auto name inference > do not overwrite manual name (call) 1`] = `
"import { defineComponent } from 'vue';
import { toDisplayString as _toDisplayString } from "vue";
export default {
    name: 'Baz',
    render (_ctx, _cache, $props, $setup, $data, $options) {
        return _toDisplayString(a);
    },
    setup (__props, { expose: __expose }) {
        __expose();
        const a = 1;
        return {
            a
        };
    }
};
"
`;

exports[`SFC analyze <script> bindings > auto name inference > do not overwrite manual name (object) 1`] = `
"import { toDisplayString as _toDisplayString } from "vue";
export default {
    name: 'Baz',
    render (_ctx, _cache, $props, $setup, $data, $options) {
        return _toDisplayString(a);
    },
    setup (__props, { expose: __expose }) {
        __expose();
        const a = 1;
        return {
            a
        };
    }
};
"
`;

exports[`SFC compile <script setup> > should expose top level declarations 1`] = `
"import { xx } from './x';
let aa = 1;
const bb = 2;
function cc() {}
class dd {
}
import { x } from './x';
export default {
    __name: "anonymous",
    setup (__props, { expose: __expose }) {
        __expose();
        let a = 1;
        const b = 2;
        function c() {}
        class d {
        }
        return {
            get xx () {
                return xx;
            },
            get aa () {
                return aa;
            },
            set aa (v){
                aa = v;
            },
            bb,
            cc,
            dd,
            get x () {
                return x;
            },
            get a () {
                return a;
            },
            set a (v){
                a = v;
            },
            b,
            c,
            d
        };
    }
};
"
`;

exports[`SFC compile <script setup> > should expose top level declarations w/ ts 1`] = `
"import { xx } from './x';
let aa = 1;
const bb = 2;
function cc() {}
class dd {
}
import { x } from './x';
import { defineComponent as _defineComponent } from "vue";
export default _defineComponent({
    __name: "anonymous",
    setup (__props, { expose: __expose }) {
        __expose();
        let a = 1;
        const b = 2;
        function c() {}
        class d {
        }
        return {
            get aa () {
                return aa;
            },
            set aa (v){
                aa = v;
            },
            bb,
            cc,
            dd,
            get a () {
                return a;
            },
            set a (v){
                a = v;
            },
            b,
            c,
            d
        };
    }
});
"
`;

exports[`SFC genDefaultAs > <script setup> only 1`] = `
"const _sfc_ = {
    __name: "anonymous",
    setup (__props, { expose: __expose }) {
        __expose();
        const a = 1;
        return {
            a
        };
    }
};
"
`;

exports[`SFC genDefaultAs > <script setup> only w/ ts 1`] = `
"import { defineComponent as _defineComponent } from "vue";
const _sfc_ = _defineComponent({
    __name: "anonymous",
    setup (__props, { expose: __expose }) {
        __expose();
        const a = 1;
        return {
            a
        };
    }
});
"
`;

exports[`SFC genDefaultAs > <script> + <script setup> 1`] = `
"const _sfc_ = {
    __name: "anonymous",
    setup (__props, { expose: __expose }) {
        __expose();
        const a = 1;
        return {
            a
        };
    }
};
"
`;

exports[`SFC genDefaultAs > <script> + <script setup> w/ ts 1`] = `
"import { defineComponent as _defineComponent } from "vue";
const _sfc_ = _defineComponent({
    __name: "anonymous",
    setup (__props, { expose: __expose }) {
        __expose();
        const a = 1;
        return {
            a
        };
    }
});
"
`;

exports[`SFC genDefaultAs > normal <script> only 1`] = `
"const _sfc_ = {
    __name: "anonymous"
};
"
`;

exports[`SFC genDefaultAs > normal <script> w/ cssVars 1`] = `
"const _sfc_ = {
    __name: "anonymous"
};
"
`;
